ome/omero-figure
# Rois from omero2

Previously opened as https://github.com/ome/omero-figure/pull/165
Ported to new code layout by squashing commits etc.

This allows loading of ROIs from OMERO to add to figure panels as new Shapes.

To test:
- In figure, pick a multi-plane image (also good to test with single plane image)
- In the Info tab, open the image with iViewer and add some ROIs (if not already on the image). At least 1 of each type of shape and including an ROI with many shapes (more than 10).
- Bonus: include rotated Ellipses (need to use Insight).
- Save ROIs, go back to OMERO.figure, and in the right tab 'ROIs' click "Draw".
- Test that drawing shapes manually works OK as before.
- Click "Load ROIs from OMERO" - (NB: this should be disabled for images with NO ROIs).
- Should see a list of ROIs / shapes, similar to view in webclient with ROIs expandable to show Shapes.
- Expanding ROIs with many shapes will not render image for more than 10 at a time (to avoid loading too many image planes at once).
- Hovering over each Shape row displays it on the image. Clicking on row sets the Z & T of the image to the Z & T of that shape. Clicking "Add" will add the Shape to the image.
- Add Arrow, Line, Ellipse and Rectangle. Check they look same as in iViewer.
- Click OK to close dialog and save shapes to image panel.
- Open the "Draw" dialog for a different image to check the dialog is re-rendered correctly.

![screen shot 2017-02-01 at 14 52 03](https://cloud.githubusercontent.com/assets/900055/22511597/1a965310-e88e-11e6-88dc-93662fdf07b9.png)


Follow up ideas:
- Maybe option to filter shapes by current plane?
- When Ellipses are added, x, y, rx, ry are recalculated from handles (to handle transforms) but this means that they are not detected as duplicates
- Maybe add Z & T sliders?
- Maybe show shapes on viewer when mouseover area!? - click to add?!
- Do we allow shape to be added if previously added & moved? Don't want duplicate IDs!
